src/bidiMapper/domains/context/logHandler.ts

```diff
@@ -0,0 +1,241 @@
+/**
+ * Copyright 2022 Google LLC.
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Protocol } from 'devtools-protocol';
+import { CdpClient } from '../../../cdp';
+import { CommonDataTypes, Log, Script } from '../../bidiProtocolTypes';
+import { IBidiServer } from '../../utils/bidiServer';
+import { ContextCommon } from './contextCommon';
+
+export class LogHandler extends ContextCommon {
+  private specifiers = ['%s', '%d', '%i', '%f', '%o', '%O', '%c'];
+
+  private constructor(
+    private _contextId: string,
+    _cdpClient: CdpClient,
+    private _bidiServer: IBidiServer,
+    EVALUATOR_SCRIPT: string
+  ) {
+    super(_cdpClient, EVALUATOR_SCRIPT);
+  }
+
+  public static async create(
+    contextId: string,
+    cdpClient: CdpClient,
+    bidiServer: IBidiServer,
+    EVALUATOR_SCRIPT: string
+  ) {
+    const logHandler = new LogHandler(
+      contextId,
+      cdpClient,
+      bidiServer,
+      EVALUATOR_SCRIPT
+    );
+    await logHandler._initialize();
+    return logHandler;
+  }
+
+  private async _initialize() {
+    this._initializeEventListeners();
+  }
+
+  private _initializeEventListeners() {
+    this._initializeLogEntryAddedEventListener();
+  }
+
+  private _getLogLevel = (consoleAPIType: string): Log.LogLevel => {
+    if (consoleAPIType in ['error', 'assert']) {
+      return 'error';
+    }
+    if (consoleAPIType in ['debug', 'trace']) {
+      return 'debug';
+    }
+    if (consoleAPIType in ['warning', 'warn']) {
+      return 'warning';
+    }
+    return 'info';
+  };
+
+  private _isFormmatSpecifier(str: string): boolean {
+    return this.specifiers.some((spec) => str.includes(spec));
+  }
+
+  /**
+   * @param args input remote values to be format printed
+   * @returns parsed text of the remote values in specific format
+   */
+  private _formatter(args: CommonDataTypes.RemoteValue[]): string {
+    let output = '';
+    const argFormat = (
+      args[0] as { type: string; value: string }
```

Comment:
> @sadym-chromium if I understand correctly, your suggestion is to avoid inheritance and instead make a ContextSerializationHelper object that Context owns and passes to the LogHandler constructor so that the LogHandler can share it. Is that correct? If so, I strongly agree with this approach.

@bwalderman yes, that what I meant.

Reply:
moved the functions to contextHelper.ts and added unit tests.

Reply:
added protection for empty input.